Load index information for MySQL, PostgreSQL, and SQLite tables

mbezhanov opened the pull request:
Hi, I am opening this in relation to https://github.com/stephenafamo/bob/issues/120.

This will allow users like @majiaxin110 to range over `.Table.Indexes` in their custom `*.tpl` enabling the use of `$index.Name` variables in a similar fashion:

```
{{$table := .Table}}
{{$tAlias := .Aliases.Table $table.Key -}}
{{range $index := .Table.Indexes}}
func print_index_{{$tAlias.DownSingular}}_{{$index.Name}}() {
    fmt.Println("{{$index.Name}}")
}
{{end}}
```

Sample output:

```go
func print_index_jet_idx_jets_names() {
	fmt.Println("idx_jets_names")
}

func print_index_jet_idx_jets_pilots() {
	fmt.Println("idx_jets_pilots")
}

func print_index_jet_PRIMARY() {
	fmt.Println("PRIMARY")
}
```

I only added MySQL support to keep the PR focused and get some early feedback on the implementation. If you approve of it, I'd happily add support for PostgreSQL and SQLite as well.

stephenafamo commented:
This is very nice! Exactly what was requested in #120

I'll take a deeper look, but overall looks great 👌🏾

mbezhanov commented:
Happy to hear that! I also included PostgreSQL support.

stephenafamo commented:
The tests are failing because the new output does not match the golden files....

You should run the generation tests locally with the `-overwrite-golden` flag to fix this.

```sh
go test ./gen/bobgen-psql/driver -overwrite-golden
```

stephenafamo commented on gen/drivers/indexes.go:
```diff
@@ -0,0 +1,10 @@
+// DBIndexes lists all indexes in the database schema keyed by table name
+type DBIndexes map[string][]Index
+
+// Index represents an index in a table
+type Index struct {
+	Name    string   `yaml:"name" json:"name"`
+	Columns []string `yaml:"columns" json:"columns"`
+}
```
We should instead reuse the `Constraint` type since they have the same structure.

Perhaps it makes more sense for the type to be called `Index` or something else, but I would rather no duplication.

mbezhanov replied:
Even though both types have the same structure, an `Index` and a `Constraint` have different semantics. Indexes are automatically created for primary key columns and unique constraints, but things are different for foreign keys. MySQL (InnoDB) automatically creates indexes for them, but PostgreSQL and SQLite do not.

Take the following DDL statements, for example, executed against a PostgreSQL database:

```sql
CREATE TABLE test.parent (
   id INT GENERATED ALWAYS AS IDENTITY,
   name VARCHAR(255) NOT NULL,
   PRIMARY KEY(id)
);

CREATE TABLE test.child (
   id INT GENERATED ALWAYS AS IDENTITY,
   parent_id INT,
   name VARCHAR(255) NOT NULL,
   PRIMARY KEY(id),
   CONSTRAINT fk_parent FOREIGN KEY(parent_id) REFERENCES parent(id)
);
```

If we run the following query to list their indexes:

```sql
SELECT * FROM pg_catalog.pg_indexes WHERE schemaname = 'test';
```

We get the following result:

```text
schemaname|tablename|indexname  |tablespace|indexdef                                                       |
----------+---------+-----------+----------+---------------------------------------------------------------+
test      |parent   |parent_pkey|          |CREATE UNIQUE INDEX parent_pkey ON test.parent USING btree (id)|
test      |child    |child_pkey |          |CREATE UNIQUE INDEX child_pkey ON test.child USING btree (id)  |
```

I can certainly generalize the struct to avoid code duplication, but neither `Index` nor `Constraint` will be a semantically correct name.

stephenafamo replied:
Yes, I understand that semantically, they are different things. This is why I said perhaps it should be called something different (not Index or Constraint).

It just feels a bit redundant to have two types that are exactly the same thing.

I guess a bit of repetition can be good for the sake of correctness.

mbezhanov replied:
What do you think about something similar:

```go
type NamedColumnList struct {
	Name    string   `yaml:"name" json:"name"`
	Columns []string `yaml:"columns" json:"columns"`
}


type Constraint struct {
	NamedColumnList
}

type Index struct {
	NamedColumnList
}
```

Or would you rather completely eliminate the `Index` type?

mbezhanov replied:
This will allow the `Index` and `Constraint` types to evolve separately (if needed).

stephenafamo replied:
> This will allow the `Index` and `Constraint` types to evolve separately (if needed).

If that time comes, then we can create these wrapper types. In the meantime, we can replace every instance with `Constraint` with `NamedColumnList`.

If we really want to have distinct types now, we can do 

```go
type Constraint NamedColumnList

type Index NamedColumnList
```

This is also great since there are no methods attached to `NamedColumnList`

stephenafamo commented:
Can you also update the SQLite golden file?

mbezhanov commented:
> Can you also update the SQLite golden file?

Yes, absolutely. I'll actually go ahead and directly implement the SQLite index retrieval logic in this PR. Coming up in a little bit :slightly_smiling_face: 

Would you like me to squash everything up to this point into a single commit, or are you planning to squash when merging into main?

stephenafamo commented:
> Would you like me to squash everything up to this point into a single commit, or are you planning to squash when merging into main?

Kindly squash and push.

I do not squash when merging into main, so the commit history can get a little messy if it is not done beforehand 😬

mbezhanov commented:
Pushing an updated version rebased on the latest `main`. I've added index listing support to `bobgen-mysql`, `bobgen-psql`, and `bobgen-sqlite`. I can do `bobgen-atlas` and `bobgen-prisma` in a separate PR, unless you prefer to have them included in this one.

I introduced a `NamedColumnList` struct as discussed, but kept `Index` and `Constraint` as distinct types rather than aliases for better readability:

```go
type Constraint NamedColumnList

type Index NamedColumnList
```

I found the aliases a bit confusing during debug sessions:

![image](https://github.com/stephenafamo/bob/assets/785542/edecd032-e80a-45b4-944b-e387c2f4a59a)

Type declarations seem to be a bit more clear:

![image](https://github.com/stephenafamo/bob/assets/785542/5c52713a-9cc1-4059-aa47-2794dc8279b1)

However, if you insist on type aliases rather than type declarations, I'd be happy to change them.

Let me know if any further improvements are necessary.

stephenafamo commented:
Looks great! Thank you.

stephenafamo commented:
I've already merged this now, but I just thought that it would be nice to be able to tell apart different types of indexes (unique vs non-unique)

mbezhanov commented:
No problem. I'll add this and open another PR in the following days.
